[DDP] Add API to get model parameters in hook

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61637**

To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

rohan-varma committed:
[DDP] Add API to get model parameters in hook

To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `dict[idx -> tensor]` where `idx` is the intra bucket index of gradient tensor and thus the same index of `perParameterTensors`. The API can be used as follows to retrieve the model parameters:

```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
for grad_tensor_idx, model_param in idx_to_model_params.items():
    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `dict[idx -> tensor]` where `idx` is the intra bucket index of gradient tensor and thus the same index of `perParameterTensors`. The API can be used as follows to retrieve the model parameters:

```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
for grad_tensor_idx, model_param in idx_to_model_params.items():
    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -301,6 +301,11 @@ to apply layer-wise operations.
     A list of ``torch.Tensor``. Each tensor in the list corresponds to a parameter.
 )")
+      .def(
+          "get_grad_index_to_variable_mapping",
```
curious, is the param index sth that users must know? Will it work if we just expose one single API `bucket.parameters()` that returns references to model parameters with the `.grad` field properly populated?

mrshenli replied:
Above might only work when `grad_as_bucket_view` is enabled. Curious, when `grad_as_bucket_view=False`, how will optimizer work in the communication hook? Are we expecting users to manually populate `param.grad` before calling `optimizer.step()`? If so, will it work if we expose `bucket.parameters` and `bucket.grads` as two separate APIs, where parameters() return references to model parameters and `grads()` return bucket views?

rohan-varma replied:
Regarding bucket.parameters and then using param.grad field, that might be hard since we will be running optimizer as a `then()` callback to gradient reduction, but `param.grad` is not populated until `Reducer::finalize_backward` is called. So if we access `param.grad` it will be a stale gradient in the comm. hook.

> Curious, when grad_as_bucket_view=False, how will optimizer work in the communication hook? Are we expecting users to manually populate param.grad before calling optimizer.step()? 

In both cases, we are planning to do something like the following:
```
grad_tensors = bucket.get_per_parameter_tensors()
idx_to_param = bucket.get_idx_to_params()
for i in range(len(grad_tensors)):
    grad, model_param = grad_tensors[i], idx_to_param[i]
    functional_optimizer.step(model_param, grad)
```

For this to work we'll need functional optimizers that operate on explicit gradients which we've created a version of for the RPC framework.

mrshenli replied:
Will sth like this work:

```python
for param, grad in zip(bucket.parameters(), bucket.gradients()):
    # param is the model parameter, whose .grad field is not populated yet
    # grad is a view to the bucket offset, and it already contains the allreduced grad. 
    functional_optimizer.step(model_param, grad)
```

mrshenli replied:
Above code means that `parameters()` and `gradients()` must match in order, but it should be fine as they both belong to the same bucket. And it will also avoid engaging with param index.

SciPioneer replied:
Probably this can be a follow-up refactoring PR.

mrshenli reviewed (commented):
cc @andwgu

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -7513,3 +7513,39 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
+                # Run allreduce and wait for it to complete
+                allreduce_fut = default._allreduce_fut(None, bucket.get_tensor())
+                allreduce_fut.wait()
+                per_param_grad_tensors = bucket.get_per_parameter_tensors()
```
Would I be correct if I assume these tensors are views on the bucket?

rohan-varma replied:
Yep, that is correct. There's no copying/cloning of tensors going on here.

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -7513,3 +7513,39 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
+                # Verify mapping is correct by ensuring param.grad is equal to
+                # the grad we get using the mapping.
+                for grad_tensor_idx, model_param in idx_to_model_params.items():
+                    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```
If the above assumption is correct, how does this work? Since the test doesn't enable `grad_as_bucket_view`, I assume `param.grad` are not bucket views, and the allreduce result are not yet copied back to `param.grad`. I might miss sth, why do we expect `param.grad` and bucket views to be the same value?

rohan-varma replied:
This is very strange because we're also not calling `zero_grad` anywhere so theoretically the grad field should accumulate. Looking into it...

rohan-varma replied:
@mrshenli This appears to pass because the gradients are so close to 0 (on the order of 1e-8) that when I print them out they are indeed different, but `self.assertEqual` default tolerance is too generous marking them as equal.

If we scale up the gradients this test indeed fails. This makes sense because during backwards the grads have not yet been copied back to param.grad field. Although, the main use case here is to just verify we can get the appropriate gradient for the parameter, so we can probably just do shape check when `grad_as_bucket_view=False` and grad check when `grad_as_bucket_view=True`?

mrshenli replied:
Yep, that looks good to me.

andwgu commented on torch/csrc/distributed/c10d/comm.hpp:
```diff
@@ -51,6 +53,15 @@ class TORCH_API GradBucket {
   std::vector<at::Tensor> getPerParameterTensors() const;
 
+  // Returns mapping of gradient index in bucket to model parameter. Practically
+  // this can be used in conjunction with getPeraParameterTensors() as follows:
```
nit: `getPeraParameterTensors()` -> `getPerParameterTensors()`

andwgu commented on torch/csrc/distributed/c10d/comm.hpp:
```diff
@@ -51,6 +53,15 @@ class TORCH_API GradBucket {
+  // Returns mapping of gradient index in bucket to model parameter. Practically
+  // this can be used in conjunction with getPeraParameterTensors() as follows:
+  // auto grads = getPerParameterTensors(); // grad tensors for this bucket
+  // auto mapping = getGradIndexToVariable();
```
nit: I think perhaps `getGradIndexToVariable()` was your old version of the function name, and this should be `auto mapping = getGradIndexToVariableMapping()` now.

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `dict[idx -> tensor]` where `idx` is the intra bucket index of gradient tensor and thus the same index of `perParameterTensors`. The API can be used as follows to retrieve the model parameters:

```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
for grad_tensor_idx, model_param in idx_to_model_params.items():
    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/comm.hpp:
```diff
@@ -21,12 +21,14 @@ class TORCH_API GradBucket {
         lengths_(lengths),
-        sizes_vec_(sizes_vec) {}
+        sizes_vec_(sizes_vec),
+        bucket_idx_to_variable_(bucket_idx_to_variable) {}
```
could we just call getVariables() to return as list of variables? as the idx is the same as idx in getPerParameterTensor() anyway?

zhaojuanmao commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -7513,3 +7513,48 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
+                for i in range(len(per_param_grad_tensors)):
+                    grad = per_param_grad_tensors[i]
+                    model_param = idx_to_model_params[i]
+                    self.assertEqual(model_param.grad, grad)
```
I think the question is: grad is averaged grad, model_param.grad is old grad, why are they equal here?

rohan-varma replied:
It was an issue because the numerics were too small and tolerance was high enough to make them equal. Should be resolved now that we check for shape when grad_as_bucket_view=False.

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `dict[idx -> tensor]` where `idx` is the intra bucket index of gradient tensor and thus the same index of `perParameterTensors`. The API can be used as follows to retrieve the model parameters:

```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
for grad_tensor_idx, model_param in idx_to_model_params.items():
    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

mrshenli reviewed (commented):
(This is prior to this PR) APIs on GradientBucket is a bit verbose. Looks like they can be simplified to, 

`get_index` -> `index` 

`get_tensor` -> `buffer` 

`set_tensor` -> `set_buffer` 

`is_the_last_bucket_to_allreduce` -> `is_last`,

`get_per_parameter_tensors` -> `parameters`, 

new API -> `gradients`. 

Since CommHook is still beta and will soon become stable, shall we take this chance to do one round of discussion and revamp to those APIs for simplicity?


--------------------

Please feel free to apply the API revamp to followup PRs

rohan-varma commented:
> (This is prior to this PR) APIs on GradientBucket is a bit verbose.
> Please feel free to apply the API revamp to followup PRs

Sure, I'll send a follow up diff with those changes.

One note:

> `get_per_parameter_tensors` -> `parameters`,
> 
> new API -> `gradients`.

I think it should basically be the other way around, as `get_per_param_tensors` actually returns grads and this API returns the model params. Name is sort of confusing.

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]

SciPioneer commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -25,6 +25,7 @@ class GradBucket:
     def get_per_parameter_tensors(self) -> List[Tensor]: ...
     def is_the_last_bucket_to_allreduce(self) -> bool: ...
     def set_tensor(self, tensor: Tensor) -> None: ...
+    def get_model_params_for_bucket(self) -> List[Tensor]: ...
```
NI: I think we can just call this `get_params`, since all the data in this `GradBucket` class is for the bucket only. No need to add extra `for_bucket`.

rohan-varma replied:
I'll call it `get_model_params` for now, otherwise could be confusing with existing `get_per_parameter_tensors`, which is actually gradients, which will be renamed in a follow up diff

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -845,17 +848,53 @@ void Reducer::all_reduce_bucket(Bucket& bucket) {
+    // Grab the actual model parameter.
+    auto& variable =
+        replica.variables[bucket_index_for_variable.intra_bucket_index];
+    variables_for_bucket.emplace_back(variable);
```
Will you still access `replica.variables`?

SciPioneer commented:
> > (This is prior to this PR) APIs on GradientBucket is a bit verbose.
> > Please feel free to apply the API revamp to followup PRs
> 
> Sure, I'll send a follow up diff with those changes.
> 
> One note:
> 
> > `get_per_parameter_tensors` -> `parameters`,
> > new API -> `gradients`.
> 
> I think it should basically be the other way around, as `get_per_param_tensors` actually returns grads and this API returns the model params. Name is sort of confusing.

Correct, previously we use "tensor" to refer to the gradient data. Now we add parameters, so better to give them better names.

rohan-varma committed:
Update on "[DDP] Add API to get model parameters in hook"


To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

[ghstack-poisoned]
